mllib/src/main/scala/org/apache/spark/ml/classification/DecisionTreeClassifier.scala

```diff
@@ -210,6 +214,22 @@ class DecisionTreeClassificationModel private[ml] (
     rootNode.predictImpl(features).prediction
   }
 
+  /** @group setParam */
+  @Since("3.0.0")
+  def setLeafCol(value: String): this.type = set(leafCol, value)
+
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    transformSchema(dataset.schema, logging = true)
+
+    if ($(leafCol).nonEmpty) {
+      val leafUDF = udf { features: Vector => predictLeaf(features) }
+      super.transform(dataset)
+        .withColumn($(leafCol), leafUDF(col($(featuresCol))))
+    } else {
+      super.transform(dataset)
```

Comment:
It's trivial but I guess you could avoid calling this in two places ... call it once and either return it, or the result of it with a new column.